### ISSUE TITLE
Problems during compilation on cygwin 32bit

### ISSUE THREAD

**Issue by tomasjura** [kubo/ruby-oci8#113]
I had problems is during compiling and linking plthook_win32.c. A command `gem install ruby-oci8` was used for installation.
* stricmp function sometimes was not present in string.h
* library dbghelp was not included in the linking process

Patch:
```c
--- ext/oci8/extconf.rb.old  2015-12-16 11:09:41.030091800 +0100
+++ ext/oci8/extconf.rb 2015-12-16 11:06:31.179893500 +0100
@@ -143,6 +183,13 @@
   raise 'unsupported ruby engine: ' + RUBY_ENGINE
 end

+case RUBY_PLATFORM
+when /mswin32|cygwin|mingw32|bccwin32/
+  have_func('stricmp', 'string.h')
+  have_func('strcasecmp', 'string.h')
+  have_library('dbghelp','ImageDirectoryEntryToData',['windows.h','dbghelp.h'])
+end
+
 print "checking for plthook... "
 STDOUT.flush
 case RUBY_PLATFORM
--- ext/oci8/plthook_win32.c.old     2015-12-16 11:09:41.123693600 +0100
+++ ext/oci8/plthook_win32.c       2015-12-16 11:06:31.179893500 +0100

@@ -39,6 +39,8 @@
 #include <windows.h>
 #include <dbghelp.h>
 #include "plthook.h"
+#include <string.h>
+#include "extconf.h"

 #ifdef _MSC_VER
 #pragma comment(lib, "dbghelp.lib")
@@ -57,6 +59,12 @@
 #define SIZE_T_FMT "u"
 #endif

+#ifndef HAVE_STRICMP
+#ifdef HAVE_STRCASECMP
+#define stricmp strcasecmp
+#endif
+#endif
+
 typedef struct {
     const char *mod_name;
     const char *name;
```

**Comment by tomasjura** [comment]
One more patch to compile on cygwin.
```c
--- ext/oci8/oraconf.rb.old     2015-12-16 12:15:04.256368800 +0100
+++ ext/oci8/oraconf.rb 2015-12-16 12:15:07.158024600 +0100
@@ -713,7 +713,7 @@
           f.puts("EXPORTS")
           open("|nm #{lib_dir}/MSVC/OCI.LIB") do |r|
             while line = r.gets
-              f.puts($1) if regex =~ line and oci_funcs.include?($1)
+              f.puts($1) if regex =~ line
             end
           end
         end
```

**Comment by kubo** [comment]
Thank you!
I'll check it later.

**Comment by tomasjura** [comment]
A small polishment for cygwin x64
```c
--- /ext/oci8/plthook_win32.c  2015-12-16 12:01:00.621345200 +0100
+++ /ext/oci8/plthook_win32.c 2015-12-17 12:31:38.911846200 +0100
@@ -54,7 +54,11 @@
 #endif

 #ifdef _WIN64
+#if defined(__CYGWIN__)
+#define SIZE_T_FMT "lu"
+#else
 #define SIZE_T_FMT "I64u"
+#endif
 #else
 #define SIZE_T_FMT "u"
 #endif
```

**Comment by kubo** [comment]
Thanks!
I committed 936071ca28f4eb620d58116e523b4ac2109526fb and 677ab253fffa249590646436a3dd33b2122dbf5e just before your comment
and I was just going to close this issue.
But I looked your code and had doubt about 677ab253fffa249590646436a3dd33b2122dbf5e, which may not be proper on mingw32 x64. Your code looks proper and simpler.
I'll check it next morning (in Japan time).

**Comment by kubo** [comment]
I fixed it by 936071ca28f4eb620d58116e523b4ac2109526fb, 677ab253fffa249590646436a3dd33b2122dbf5e and 5571da4084396f9c8567951de8ca4f2b5cb783a4.

Thanks again.

5571da4084396f9c8567951de8ca4f2b5cb783a4 uses `_LP64` to decide the printf format of size_t on cygwin 64 because it is safer when the code snippet is reused. FIY `_WIN64` is not defined without including `windows.h` on cygwin 64 though it is included there.
